[2021-08-04] Add post content negotiation w/ microprofile (#836)
oktadev/okta-blog: _source/_posts/2021-08-04-microprofile-content-negotaion.adoc

---
layout: blog_post
title: "Content Negotiation with a Java MicroProfile Application"
author: brian-demers

> by: advocate

Author: ```suggestion
by: something else
```

== Prerequisites
- https://adoptopenjdk.net/[Java 11+]
- https://maven.apache.org/download.cgi[Apache Maven]
- https://httpie.io/docs#installation[HTTPie]

If you want to skip ahead, the source for this example is on https://github.com/oktadev/quarkus-content-negotiation-example[GitHub].

== What is MicroProfile?

Around 2016, the Java EE specifications had started to stagnate and suffered from long release cycles; as a result, newer web services APIs were falling behind. Created to spur innovation, the new https://microprofile.io/[MicroProfile] project addressed the need to manage the existing Java JAX-RS, CDI, and JSON-P APIs, along with other new APIs for the evolving micro-services world.

Fast forward to today, Java EE has moved to https://jakarta.ee/[Jakarta EE] under the https://www.eclipse.org/org/foundation/[Eclipse Foundation] and many MicroProfile projects have matured into Jakarta EE projects.

TIP: Some projects change root packages as they move between projects. Anything starting with `javax.\*` may have moved to `jakarta.*` or `org.eclipse.*`.

== What is Content Negotiation?

> In server-driven content negotiation, the client makes a request to a server with instructions on the type of response it can handle. When possible, the server responds with the appropriate  format or returns a `406` or `415` status code.

Reviewer: ```suggestion
by: Havi 
remove extra space btwn appropriate & format : 
 the appropriate  format or returns a `406` or `415` status code.-->  the appropriate format ... `415` status code
```

> In server-driven content negotiation, the client makes a request to a server with instructions on the type of response it can handle. When possible, the server responds with the appropriate  format or returns a `406` or `415` status code.

Author: ```suggestion
In server-driven content negotiation, the client makes a request to a server with instructions on the type of response it can handle. When possible, the server responds with the appropriate format or returns a `406` or `415` status code.
```

At a high level, the conversation looks like this:

====
**Client:** +
Hey Server! +
I'd like to look at `https://api.example.com/user/123`. +
I need your response in JSON. +
====

====
[.text-right]
**Server:** +
No Problem Client!  Here is that response...
====

Another more verbose example might be:

====
**Client:** +
Hey Server!  I need `https://api.example.com/user/123`, preferably in JSON! +
But, 😩 I'll take XML if that's all you have. +
I also need the info in English or French. +
Oh, please zip up the contents too. +

[source,http]
----
GET /user/123 HTTP/1.1
Accept: application/json,application/xml;q=0.9
Accept-Encoding: gzip
Accept-Language: en,fr
Host: api.example.com
User-Agent: Client/2.0
----

====

====
[.text-right]
**Server:** +
Hey Client! +
All I have is XML (sorry about that), the response is in English, +
and I was able to zip it; here you go...

[source,http,indent=40]
----
HTTP/1.1 200 OK
Content-Type: application/xml
Content-Encoding: gzip
Content-Language: en

<user id="123">
  ...
</user>
----
====

I'd like to take this opportunity to apologize for using XML in this example 🤪

> NOTE: *Agent-driven* content negotiation works differently. In this case,the client must already be aware of the server's capabilities or must determine them by making a request to the server. This communication is not standardized.

Reviewer: ```suggestion
By: Havi 
add space after comma
In this case,the client --> In this case, the client
```

> NOTE: *Agent-driven* content negotiation works differently. In this case,the client must already be aware of the server's capabilities or must determine them by making a request to the server. This communication is not standardized.

Author: ```suggestion
NOTE: *Agent-driven* content negotiation works differently. In this case, the client must already be aware of the server's capabilities or must determine them by making a request to the server. This communication is not standardized.
```

The server's response contains headers to let the client know which options were selected.

- `Content-Type` - The media type contained in the response.
- `Content-Encoding` - The compression algorithm used.
- `Content-Language` - The language of the response.

When I examine the request my browser made to display this blog post, it looks something like this:

|=======
|Accept |text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
|Accept-Encoding |gzip, deflate, br
|Accept-Language |en-US,en;q=0.9
|=======

My browser is telling the server: I'll take anything, but here's an ordered list of what I prefer:

1. HTML, XHTML, AVIF, WebP, or APNG
2. XML or a signed exchange
3. Anything else.

The browser supports `gzip`, `deflate`, and `br` compression. Finally, my language preference is US English (as in color spelled without a "u") or any English.

IMPORTANT: There is a downside. These headers can also be used for https://coveryourtracks.eff.org/[browser tracking and finger printing].

Enough primer, let's get on to the code!

== Create a MicroProfile Project With Quarkus

> Most of the code below should be vendor agnostic, except where noted.Try out the example with your favorite https://start.microprofile.io/[MicroProfile vendor], and let me know how it went in the comments!

Reviewer: ```suggestion
Most of the code below should be vendor agnostic, except where noted. Try out the example with your favorite https://start.microprofile.io/[MicroProfile vendor], and let me know how it went in the comments!
```
By: Havi 
except where noted.Try out --> except where noted. Try out

    -DprojectArtifactId=content-negotation \
    -DclassName="com.example.DiceResource" \
    -Dextensions="resteasy-jsonb"
    -Dpath="/roll"
cd content-negotation
----

If you haven't guessed, this example is going to evaluate dice notation. The https://github.com/diceroll-dev/dice-parser[Dice Notation Parser] library will evaluate dice expressions such as `2d8+2` and return the results. Open up your `pom.xml` file and add the `dice-parser` dependency:

[source,xml]
----
<dependency>
    <groupId>dev.diceroll</groupId>
    <artifactId>dice-parser</artifactId>
    <version>0.1.0</version>
</dependency>
----

Update the `DiceResource` class in `src/main/java/com/example/resources` to add a `GET` method:

[source,java]
----
package com.example.resources;

import dev.diceroll.parser.ParseException;
import dev.diceroll.parser.ResultTree;

import javax.ws.rs.*;
import javax.ws.rs.core.*; // wildcard for brevity

import static dev.diceroll.parser.Dice.detailedRoll;

@Path("/roll")
public class DiceResource  {

    @GET
    public ResultTree rollObject(@QueryParam("dice") String dice) throws ParseException {
        return detailedRoll(dice);
    }
}
----

> Start the server either from your favorite IDE or, from the command line with:

Reviewer: ```suggestion
Start the server either from your favorite IDE or from the command line with:
```
remove misplaced comma

            .languages(Locale.ENGLISH, Locale.GERMAN) // <.>
            .build();

    @Override
    public void filter(ContainerRequestContext requestContext) {
        Variant variant = requestContext.getRequest().selectVariant(VARIANTS); // <.>

        if (variant == null) { // <.>
            // Error, respond with 406
            requestContext.abortWith(Response.notAcceptable(VARIANTS).build());
        } else {
            // keep the resolved lang around for the response
            requestContext.setProperty(LANG, variant.getLanguageString()); // <.>
        }
    }

    @Override
    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
        String lang = (String) requestContext.getProperty(LANG);
        responseContext.getHeaders().putSingle(HttpHeaders.CONTENT_LANGUAGE, lang); // <.>
    }
}
----

<.> Define the supported languages.
<.> Select the `Variant` based on the request.
<.> If a compatible `Variant` is not found, return a `406`.
<.> Add the resolved language string to the request context.
<.> Set the `Content-Language` response header.
====

> NOTE: Having spent years dealing with the Servlet API, I often make incorrect assumptions about how the JAX-RS API is structured. With JAX-RS the request and response filtering are different interfaces.

Reviewer: ```suggestion
NOTE: With JAX-RS, the request and response filtering are different interfaces.
```
Insert comma after JAX-RS for clarity.

> NOTE: Having spent years dealing with the Servlet API, I often make incorrect assumptions about how the JAX-RS API is structured. With JAX-RS the request and response filtering are different interfaces.

Reviewer: ```suggestion
NOTE: Having spent years dealing with the Servlet API, I often make incorrect assumptions about how the JAX-RS API is structured. With JAX-RS, the request and response filtering are different interfaces.
```
Insert comma after JAX-RS for clarity

This solution works with any endpoint; test out the original `/roll` endpoint again. Dice notation allows for basic math, too; try to roll `2d8+1` (two eight-sided dice plus one).

====
[source,sh]
----
http :8080/roll dice==2d8+1 "Accept-Language: en"
----

[source,http]
----
HTTP/1.1 200 OK
Content-Language: en
Content-Type: application/json
Vary: Accept-Language
content-encoding: gzip
content-length: 154

{
    "expression": {
        "left": {
            "numberOfDice": 2,
            "numberOfFaces": 8
        },
        "operation": "ADD",
        "right": {
            "value": 1
        }
    },
    "results": [
    ...
    ],
    "value": 8
}
----
====

So far, so good!

== Different Responses for Different Clients

> Up until now, the simple `/roll` endpoint has been returning JSON, which is excellent if you are building a client that accepts JSON, but it's not very user-friendly for text-based clients like HTTPie or curl. Two changes are needed to support a plain text response, annotate endpoint with an `@Produces` annotation and add a `MessageBodyWriter` that converts the dice parser's `ResultTree` object into text.

Reviewer: ```suggestion
Up until now, the simple `/roll` endpoint has been returning JSON, which is excellent if you are building a client that accepts JSON, but it's not very user-friendly for text-based clients like HTTPie or curl. Two changes are needed to support a plain text response: Annotate the endpoint with an `@Produces` annotation and add a `MessageBodyWriter` to convert the dice parser's `ResultTree` object into text.
```
Fix punctuation for "Two changes..." and adjust spacing

Add the following `@Produces` annotation to `DiceResource`, to indicate it supports both `text/plain` and `application/json`:

[source,java]
----
@Path("/roll")
@Produces({MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON })
public class DiceResource  {
...
----

If you make a request to the `/roll` endpoint now, you will get an ugly response due to a call to the `toString()` method on `ResultTree`. 🤢

> To fix this, create a new class `TextDiceTreeMessageBodyWriter.java` that implements `MessageBodyWriter<ResultTree>`, which render something a little easier on the eyes:

Reviewer: ```suggestion
To fix this, create a new class `TextDiceTreeMessageBodyWriter.java` that implements `MessageBodyWriter<ResultTree>`, which renders something a little easier on the eyes:
```
correct verb form - render -> renders

import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.ext.MessageBodyWriter;
import javax.ws.rs.ext.Provider;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;

@Provider
@Produces(MediaType.TEXT_PLAIN) // <.>
public class TextDiceTreeMessageBodyWriter implements MessageBodyWriter<ResultTree> {

    @Override
    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
        return type == ResultTree.class; // <.>
    }

    @Override
    public void writeTo(ResultTree resultTree,
                        Class<?> type, Type genericType,
                        Annotation[] annotations,
                        MediaType mediaType,
                        MultivaluedMap<String, Object> httpHeaders,
                        OutputStream entityStream) throws IOException, WebApplicationException {

        String result = Dice.debug(resultTree); // <.>
        entityStream.write(result.getBytes(StandardCharsets.UTF_8)); // <.>
    }
}
----

> <.> Add a `Produces` annotation to the class to mark the `MediaType's it supports.

Reviewer: ```suggestion
<.> Add a `Produces` annotation to the class to mark the `MediaType`s it supports.
```
correct backtick punctuation

Author: I just bumped into this, the trailing `s` causes the code block to not render.  (accepting changes but also dropping the trailing `s`)

<.> This example keeps things simple and only supports the `ResultTree` class.
<.> Calling `Dice.debug` will turn the result into text.
<.> Write the string to the response stream.
====

Time to test it out! Call `/roll` again roll `3d10k2+1` (roll three eight-sided dice and keep the highest two, then add one to the result):

====
[source,sh]
----
http :8080/roll dice==3d10k2 "Accept: text/plain"
----

[source,http]
----
HTTP/1.1 200 OK
Content-Language: en
Content-Type: text/plain;charset=UTF-8
Vary: Accept-Language
content-encoding: gzip
content-length: 69

3d10k2 + 1 = 8
--3d10k2 = 7
----d10 = 4
----d10 = 2
----d10 = 3
--1 = 1
----
====

Much better! Change the `Accept` header value to `application/json` to see for yourself.

== Bonus: User Agent for Content Negotiation

> Using the User-Agent header for content negotiation might not be your first choice. Still, it works great for a few use cases, for example, when dealing with legacy clients if they don't _correctly_ handle adding new fields to a JSON response object.

Reviewer: ```suggestion
Using the User-Agent header for content negotiation might not be your first choice. Still, it works great for a few use cases, for example when dealing with legacy clients that don't _correctly_ handle adding new fields to a JSON response object.
```
removed comma & "if they" for better flow.

TIP: If you are using Jackson to unmarshal JSON consider adding the `@JsonIgnoreProperties(ignoreUnknown = true)` annotation to your client code to handle additive changes in REST APIs gracefully.

> One of my favorite examples of using the User-Agent for content negotiation is adding descriptive help to a REST server. I learned this trick from the [Spring Initializer project](https://github.com/spring-io/start.spring.io), which returns different results if you use HTTPie or curl: `http https://start.spring.io`.

Reviewer: ```suggestion
One of my favorite examples of using the User-Agent for content negotiation is adding descriptive help to a REST server. I learned this trick from the [Spring Initializer project]https://github.com/spring-io/start.spring.io, which returns different results if you use HTTPie or curl: `http https://start.spring.io`.
```
Borked hyperlink: removed parens

To detect the user agent, add an enum that contains basic user-agent parsing logic.

====
[source,java]
----
package com.example;

import java.util.Arrays;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum Agents {

    CURL("curl"),
    HTTPIE("HTTPie");

    final private String agentName;
    final private static Pattern TOOL_REGEX = Pattern.compile("([^\\/]*)\\/([^ ]*).*"); // <.>

    Agents(String agentName) {
        this.agentName = agentName;
    }

    public static Agents parse(String userAgent) { // <.>
        Matcher matcher = TOOL_REGEX.matcher(userAgent);
        String name = (matcher.matches()) ? matcher.group(1) : null;
        return Arrays.stream(Agents.values())
                .filter(agent -> agent.agentName.equals(name))
                .findFirst()
                .orElse(null);
    }
}
----

<.> This regex looks for `{name}/{version}`
<.>  Calling `Agents.parse(userAgent)` will return an enum associated with Curl, HTTPie, or `null`.
====

WARNING: User-Agent parsing is a much more complex topic; the above code works well because we are looking for a few specific clients.  This is not a general-purpose solution.

> Now that I have logic to figure out the user-agent, I'll create a helper class to wrap the logic of returning localized help text.  Create a new class `Help.java`:

Reviewer: ```suggestion
Now that I have logic to figure out the user-agent, I'll create a helper class to wrap the logic of returning localized help text. Create a new class `Help.java`:
```
removed extra spacing before "Create..."

====
[source,java]
----
package com.example;

import javax.ws.rs.core.UriInfo;
import java.text.MessageFormat;
import java.util.Locale;
import java.util.Map;
import java.util.ResourceBundle;

public class Help {

    final static private Map<Agents, String> AGENT_HELP_MAP = Map.of(
            Agents.CURL, "help.curl",
            Agents.HTTPIE, "help.httpie"
    );

    public static String getHelp(String userAgent, UriInfo uriInfo, Locale locale) {

        String url = uriInfo.getBaseUri().resolve("/").toString(); // <.>

        Agents agent = Agents.parse(userAgent); // <.>

        // look up the help key
        String helpKey = AGENT_HELP_MAP.getOrDefault(agent, "help.generic");

        // Resource Bundle lookup/formatting
        ResourceBundle resourceBundle = ResourceBundle.getBundle("messages", locale);
        MessageFormat formatter = new MessageFormat(resourceBundle.getString(helpKey), locale);
        return formatter.format(new Object[] { url });
    }
}
----

<.> Resolve the base URL of the request `UriInfo`.
<.> Parse the user-agent header and return an `Agents` enum.

> <.> Lookup the help text key, or default to `help.generic`.

Reviewer: ```suggestion
<.> Look up the help text key, or default to `help.generic`.
```

Lookup-> Look up

                request.selectVariant(LanguageFilter.VARIANTS).getLanguage()); // <.>
    }
}
----
<.> The User-Agent request header.
<.> The `Request` object (needed to select the correct language `Variant`).
<.> The URI of the request.
<.> Pass everything into `Help.getHelp()` to return the localized help message.
====

The default Quarkus application adds a static `index.html` file to use as the *root* resource; delete this file so the new `HelpResource` can take its place:

[source,sh]
----
rm src/main/resources/META-INF/resources/index.html
----

Put everything you learned together and make a request to `localhost:8080/`:

====
[source,sh]
----
http :8080/ "Accept-Language: en" "Accept: text/plain" "Accept-encoding: gzip"
----

[source,http]
----
HTTP/1.1 200 OK
Content-Language: en
Content-Type: text/plain;charset=UTF-8
Vary: Accept-Language
content-encoding: gzip
content-length: 251

Welcome to the Dice Parser!!

Roll dice by making a request:
    http http://localhost:8080/roll?dice=2d6 "Accept: text/plain"

Or get the result in JSON:
    http http://localhost:8080/roll?dice=2d6 "Accept: application/json"
----
====

== Learn More About Java Web Services

> In this post, you have learned how to build a simple MicroProfile application using Qurakus and the JAX-RS API to take advantage of content negotiation.

Reviewer: ```suggestion
In this post, you have learned how to build a simple MicroProfile application using Quarkus and the JAX-RS API to take advantage of content negotiation.
```
Quarkus NOT Qurakus :)

Author: Good catch!!
